File: ios/DPIPWidgets/CurrentWeatherWidgetSnapshot.swift
import Foundation
import SwiftUI

enum CurrentWeatherWidgetCondition: String, Decodable, Sendable {
    case clear
    case cloudy
    case overcast
    case rain
    case thunderstorm
    case snow
    case fog
    case unknown

    var localizedDisplayName: LocalizedStringKey {
        LocalizedStringKey(displayNameLocalizationKey)
    }

    var displayNameLocalizationKey: String {
        switch self {
        case .clear:
            return "weather.clear"
        case .cloudy:
            return "weather.cloudy"
        case .overcast:
            return "weather.overcast"
        case .rain:
            return "weather.rain"
        case .thunderstorm:
            return "weather.thunderstorm"
        case .snow:
            return "weather.snow"
        case .fog:
            return "weather.fog"
        case .unknown:
            return "weather.unknown"
        }
    }

    func systemImageName(isNight: Bool) -> String {
        switch self {
        case .clear:
            return isNight
                ? "moon.stars.fill"
                : "sun.max.fill"

        case .cloudy:
            return isNight
                ? "cloud.moon.fill"
                : "cloud.sun.fill"

        case .overcast:
            return "cloud.fill"

        case .rain:
            return "cloud.rain.fill"

        case .thunderstorm:
            return "cloud.bolt.rain.fill"

        case .snow:
            return "cloud.snow.fill"

        case .fog:
            return "cloud.fog.fill"

        case .unknown:
            return "cloud.fill"
        }
    }
}

struct CurrentWeatherWidgetSnapshot: Codable, Sendable {
    let schemaVersion: Int
    let sourceIdentifier: String?

    let regionCode: String
    let regionName: String

    let observationTime: Int

    let stationName: String

    let weather: String
    let weatherCode: Int
    let condition: CurrentWeatherWidgetCondition
    let isNight: Bool

    /// The one next solar transition carried by this snapshot.
    let nextDayNightTransitionTime: Int

    /// Calibrated/server time minus device time, in milliseconds.
    let calibratedTimeOffsetMilliseconds: Int

    let temperature: Double?
    let humidity: Int?
    let rain: Double?

    private enum CodingKeys: String, CodingKey {
        case schemaVersion
        case sourceIdentifier
        case regionCode
        case regionName
        case observationTime
        case stationName
        case weather
        case weatherCode
        case condition
        case isNight
        case nextDayNightTransitionTime
        case calibratedTimeOffsetMilliseconds
        case temperature
        case humidity
        case rain
    }

    init(
        schemaVersion: Int,
        sourceIdentifier: String? = nil,
        regionCode: String,
        regionName: String,
        observationTime: Int,
        stationName: String,
        weather: String,
        weatherCode: Int,
        condition: CurrentWeatherWidgetCondition,
        isNight: Bool,
        nextDayNightTransitionTime: Int,
        calibratedTimeOffsetMilliseconds: Int,
        temperature: Double?,
        humidity: Int?,
        rain: Double?
    ) {
        self.schemaVersion = schemaVersion
        self.sourceIdentifier = sourceIdentifier
        self.regionCode = regionCode
        self.regionName = regionName
        self.observationTime = observationTime
        self.stationName = stationName
        self.weather = weather
        self.weatherCode = weatherCode
        self.condition = condition
        self.isNight = isNight
        self.nextDayNightTransitionTime = nextDayNightTransitionTime
        self.calibratedTimeOffsetMilliseconds =
            calibratedTimeOffsetMilliseconds
        self.temperature = temperature
        self.humidity = humidity
        self.rain = rain
    }
    init(from decoder: Decoder) throws {
        let container = try decoder.container(
            keyedBy: CodingKeys.self
        )

        schemaVersion = try container.decode(
            Int.self,
            forKey: .schemaVersion
        )

        guard (1...5).contains(schemaVersion) else {
            throw DecodingError.dataCorruptedError(
                forKey: .schemaVersion,
                in: container,
                debugDescription:
                    "Unsupported current-weather snapshot schema version."
            )
        }

        if schemaVersion >= 5 {
            sourceIdentifier = try container.decode(
                String.self,
                forKey: .sourceIdentifier
            )
        } else {
            sourceIdentifier = nil
        }

        regionCode = try container.decode(
            String.self,
            forKey: .regionCode
        )

        regionName = try container.decode(
            String.self,
            forKey: .regionName
        )

        observationTime = try container.decode(
            Int.self,
            forKey: .observationTime
        )

        stationName = try container.decode(
            String.self,
            forKey: .stationName
        )

        weather = try container.decode(
            String.self,
            forKey: .weather
        )

        weatherCode = try container.decode(
            Int.self,
            forKey: .weatherCode
        )

        let conditionRawValue = try container.decodeIfPresent(
            String.self,
            forKey: .condition
        )

        condition = conditionRawValue
            .flatMap(CurrentWeatherWidgetCondition.init(rawValue:))
            ?? .unknown

        isNight = try container.decodeIfPresent(
            Bool.self,
            forKey: .isNight
        ) ?? false

        nextDayNightTransitionTime = try container.decodeIfPresent(
            Int.self,
            forKey: .nextDayNightTransitionTime
        ) ?? 0

        if schemaVersion >= 4 {
            calibratedTimeOffsetMilliseconds = try container.decode(
                Int.self,
                forKey: .calibratedTimeOffsetMilliseconds
            )
        } else {
            calibratedTimeOffsetMilliseconds = try container.decodeIfPresent(
                Int.self,
                forKey: .calibratedTimeOffsetMilliseconds
            ) ?? 0
        }

        temperature = try container.decodeIfPresent(
            Double.self,
            forKey: .temperature
        )

        humidity = try container.decodeIfPresent(
            Int.self,
            forKey: .humidity
        )

        rain = try container.decodeIfPresent(
            Double.self,
            forKey: .rain
        )
    }

    func encode(to encoder: Encoder) throws {
        guard schemaVersion == 5 else {
            throw EncodingError.invalidValue(
                schemaVersion,
                .init(
                    codingPath: encoder.codingPath,
                    debugDescription:
                        "Only current-weather snapshot schema version 5 can be encoded."
                )
            )
        }

        guard let sourceIdentifier else {
            throw EncodingError.invalidValue(
                sourceIdentifier as Any,
                .init(
                    codingPath: encoder.codingPath,
                    debugDescription:
                        "Schema version 5 requires a source identifier."
                )
            )
        }

        guard let address = CurrentWeatherSnapshotAddress(
            sourceIdentifier: sourceIdentifier
        ) else {
            throw EncodingError.invalidValue(
                sourceIdentifier,
                .init(
                    codingPath: encoder.codingPath,
                    debugDescription:
                        "Invalid current-weather snapshot source identifier."
                )
            )
        }

        guard CurrentWeatherSnapshotAddress(
            sourceIdentifier: "region:\(regionCode)"
        ) != nil else {
            throw EncodingError.invalidValue(
                regionCode,
                .init(
                    codingPath: encoder.codingPath,
                    debugDescription:
                        "Current-weather snapshot region code must be exactly three ASCII digits."
                )
            )
        }

        if case let .saved(addressRegionCode) = address {
            guard addressRegionCode == regionCode else {
                throw EncodingError.invalidValue(
                    regionCode,
                    .init(
                        codingPath: encoder.codingPath,
                        debugDescription:
                            "Saved snapshot source identifier and payload region code must match."
                    )
                )
            }
        }

        var container = encoder.container(
            keyedBy: CodingKeys.self
        )

        try container.encode(
            schemaVersion,
            forKey: .schemaVersion
        )
        try container.encode(
            sourceIdentifier,
            forKey: .sourceIdentifier
        )
        try container.encode(
            regionCode,
            forKey: .regionCode
        )
        try container.encode(
            regionName,
            forKey: .regionName
        )
        try container.encode(
            observationTime,
            forKey: .observationTime
        )
        try container.encode(
            stationName,
            forKey: .stationName
        )
        try container.encode(
            weather,
            forKey: .weather
        )
        try container.encode(
            weatherCode,
            forKey: .weatherCode
        )
        try container.encode(
            condition.rawValue,
            forKey: .condition
        )
        try container.encode(
            isNight,
            forKey: .isNight
        )
        try container.encode(
            nextDayNightTransitionTime,
            forKey: .nextDayNightTransitionTime
        )
        try container.encode(
            calibratedTimeOffsetMilliseconds,
            forKey: .calibratedTimeOffsetMilliseconds
        )

        if let temperature {
            try container.encode(
                temperature,
                forKey: .temperature
            )
        } else {
            try container.encodeNil(
                forKey: .temperature
            )
        }

        if let humidity {
            try container.encode(
                humidity,
                forKey: .humidity
            )
        } else {
            try container.encodeNil(
                forKey: .humidity
            )
        }

        if let rain {
            try container.encode(
                rain,
                forKey: .rain
            )
        } else {
            try container.encodeNil(
                forKey: .rain
            )
        }
    }
}

File: lib/features/weather/domain/current_weather_widget_snapshot.dart
import 'package:dpip/core/weather/weather_code.dart';
import 'package:dpip/features/weather/domain/apparent_temperature.dart';
import 'package:dpip/features/weather/domain/weather_realtime.dart';

enum CurrentWeatherWidgetCondition {
  clear,
  cloudy,
  overcast,
  rain,
  thunderstorm,
  snow,
  fog,
  unknown,
}

final class CurrentWeatherWidgetSnapshot {
  const CurrentWeatherWidgetSnapshot({
    this.schemaVersion = 7,
    required this.sourceIdentifier,
    required this.regionCode,
    required this.regionName,
    required this.observationTime,
    required this.stationName,
    required this.weather,
    required this.weatherCode,
    required this.condition,
    required this.isNight,
    required this.nextDayNightTransitionTime,
    required this.calibratedTimeOffsetMilliseconds,
    this.temperature,
    this.humidity,
    this.rain,
    this.windDirection,
    this.windSpeed,
    this.apparentTemperature,
  });

  final int schemaVersion;

  final String sourceIdentifier;

  final String regionCode;
  final String regionName;

  final int observationTime;

  final String stationName;

  final String weather;
  final int weatherCode;
  final CurrentWeatherWidgetCondition condition;
  final bool isNight;

  /// The next solar transition known when this snapshot was written.
  ///
  /// This is intentionally one transition, not an indefinite solar schedule.
  final int nextDayNightTransitionTime;

  /// Calibrated/server time minus device time, in milliseconds.
  ///
  /// Swift adds this value to a device-clock `Date` to reconstruct calibrated
  /// time, and subtracts it from calibrated deadlines for WidgetKit scheduling.
  final int calibratedTimeOffsetMilliseconds;

  final double? temperature;
  final int? humidity;
  final double? rain;
  final String? windDirection;
  final double? windSpeed;
  final double? apparentTemperature;

  Map<String, Object?> toJson() {
    return {
      'schemaVersion': schemaVersion,
      'sourceIdentifier': sourceIdentifier,
      'regionCode': regionCode,
      'regionName': regionName,
      'observationTime': observationTime,
      'stationName': stationName,
      'weather': weather,
      'weatherCode': weatherCode,
      'condition': condition.name,
      'isNight': isNight,
      'nextDayNightTransitionTime': nextDayNightTransitionTime,
      'calibratedTimeOffsetMilliseconds': calibratedTimeOffsetMilliseconds,
      'temperature': temperature,
      'humidity': humidity,
      'rain': rain,
      'windDirection': windDirection,
      'windSpeed': windSpeed,
      'apparentTemperature': apparentTemperature,
    };
  }
}

CurrentWeatherWidgetCondition currentWeatherWidgetCondition(int code) {
  return switch (weatherConditionForCode(code)) {
    WeatherCondition.clear => CurrentWeatherWidgetCondition.clear,
    WeatherCondition.cloudy => CurrentWeatherWidgetCondition.cloudy,
    WeatherCondition.overcast => CurrentWeatherWidgetCondition.overcast,
    WeatherCondition.rain => CurrentWeatherWidgetCondition.rain,
    WeatherCondition.thunderstorm => CurrentWeatherWidgetCondition.thunderstorm,
    WeatherCondition.snow => CurrentWeatherWidgetCondition.snow,
    WeatherCondition.fog => CurrentWeatherWidgetCondition.fog,
    WeatherCondition.unknown => CurrentWeatherWidgetCondition.unknown,
  };
}

CurrentWeatherWidgetSnapshot createCurrentWeatherWidgetSnapshot({
  required String sourceIdentifier,
  required String regionCode,
  required String regionName,
  required WeatherRealtime weather,
  required bool isNight,
  required int nextDayNightTransitionTime,
  required int calibratedTimeOffsetMilliseconds,
}) {
  return CurrentWeatherWidgetSnapshot(
    sourceIdentifier: sourceIdentifier,
    regionCode: regionCode,
    regionName: regionName,
    observationTime: weather.time,
    stationName: weather.station.name,
    weather: weather.data.weather,
    weatherCode: weather.data.weatherCode,
    condition: currentWeatherWidgetCondition(weather.data.weatherCode),
    isNight: isNight,
    nextDayNightTransitionTime: nextDayNightTransitionTime,
    calibratedTimeOffsetMilliseconds: calibratedTimeOffsetMilliseconds,
    temperature: weather.data.temperature,
    humidity: weather.data.humidity,
    rain: weather.data.rain,
    windDirection: weather.data.wind.direction,
    windSpeed: weather.data.wind.speed,
    apparentTemperature: currentApparentTemperature(
      temperature: weather.data.temperature,
      humidity: weather.data.humidity,
      windSpeed: weather.data.wind.speed,
    ),
  );
}
